plutus-core/plutus-core/test/CostModelSafety/Spec.hs

```diff
@@ -0,0 +1,209 @@
+function which is maximally expensive, but the implementer might miss that.
+It's still possible to provide a costing function which is unrealistically
+cheap, but it would be difficult to spot that automatically.  Here we check that
+the costing functions for each builtin are nonzero at a single point, and we do
+this by running the function with a list of small arguments.  For CPU costs we
+actually check that the cost is at least 1000 ExCPU and for memory costs we
+check that the cost is strictly positive. -}
+
+module CostModelSafety.Spec (test_costModelSafety)
+where
+
+import PlutusCore (DefaultFun, DefaultUni)
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin
+import PlutusCore.Crypto.BLS12_381.G1 qualified as BLS12_381.G1
+import PlutusCore.Crypto.BLS12_381.G2 qualified as BLS12_381.G2
+import PlutusCore.Crypto.BLS12_381.Pairing qualified as BLS12_381.Pairing
+import PlutusCore.Data (Data (..))
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget (ExBudget))
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultBuiltinCostModel,
+                                                          defaultCekMachineCosts)
+import PlutusCore.Evaluation.Machine.ExBudgetStream (sumExBudgetStream)
+import PlutusCore.Evaluation.Machine.ExMemoryUsage (LiteralByteSize)
+import UntypedPlutusCore.Evaluation.Machine.Cek.CekMachineCosts (CekMachineCosts,
+                                                                 CekMachineCostsBase (..))
+
+import Data.ByteString qualified as BS
+import Data.Functor.Identity (Identity (..))
+import Data.Kind qualified as GHC (Type)
+import Data.List.Extra (enumerate)
+import Data.Text (Text)
+import Data.Word (Word8)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (assertBool, testCase)
+import Type.Reflection (TypeRep, eqTypeRep, pattern App, typeRep, (:~~:) (..))
+
+-- Machine costs
+checkBudget :: Identity ExBudget  -> IO ()
+checkBudget (Identity (ExBudget cpu mem)) = do
+  assertBool "exBudgetCPU  <= 0 in CEK machine costs" $ cpu > 0
+  assertBool "exBudgetMemory <= 0 in CEK machine costs" $ mem > 0
+
+-- Check that the machine costs are all strictly positive.  All of the fields are matched explicitly
+-- to make sure that we don't forget any new ones that get added.
+testMachineCostModel :: CekMachineCosts -> IO ()
+testMachineCostModel (
+  CekMachineCostsBase
+    cekStartupBudget
+    cekVarBudget
+    cekConstBudget
+    cekLamBudget
+    cekDelayBudget
+    cekForceBudget
+    cekApplyBudget
+    cekBuiltinBudget
+    cekConstrBudget
+    cekCaseBudget
+  ) = do
+    checkBudget cekStartupBudget
+    checkBudget cekVarBudget
+    checkBudget cekConstBudget
+    checkBudget cekLamBudget
+    checkBudget cekDelayBudget
+    checkBudget cekForceBudget
+    checkBudget cekApplyBudget
+    checkBudget cekBuiltinBudget
+    checkBudget cekConstrBudget
+    checkBudget cekCaseBudget
+
+-- Builtin costs
+
+{- Much of the code here is based on Evaluation.Spec and
```

Comment:
Could we instead just extend those tests?

Reply:
It already took me quite a long time to get what's there to work and I was running out of JIRA points so I thought I'd quit while I was ahead.  @effectfully already had an issue for improved tests (PLT-881), so I thought I'd let him do something more comprehensive later.

Reply:
I'm going to murder it all at some point and rewrite it completely with `QuickCheck` ('cause we have really good  generators for it), so let's not worry about polishing those tests.

Reply:
> I'm going to murder it all at some point and rewrite it completely with QuickCheck 

Good!  I look forward to that.